src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,257 @@
+                long currentTime = Stopwatch.GetTimestamp();
+
+                if (_connection._rttEstimator.MinRtt > TimeSpan.Zero)
+                {
+                    TimeSpan rtt = _connection._rttEstimator.MinRtt;
+                    TimeSpan dt = StopwatchTicksToTimeSpan(currentTime - _lastWindowUpdate);
+
+                    // We are detecting bursts in the amount of data consumed within a single 'dt' window update period.
+                    // The value "_deliveredBytes / dt" correlates with the bandwidth of the connection.
+                    // We need to extend the window, if the bandwidth-delay product grows over the current window size.
+                    // To enable empirical fine tuning, we apply a configurable multiplier (_windowScaleThresholdMultiplier) to the window size, which defaults to 1.0
+                    //
+                    // The condition to extend the window is:
+                    // (_deliveredBytes / dt) * rtt > _streamWindowSize * _windowScaleThresholdMultiplier
+                    //
+                    // Which is reordered into the form below, to avoid the division:
+                    if (_deliveredBytes * rtt.Ticks > _streamWindowSize * dt.Ticks * _windowScaleThresholdMultiplier)
+                    {
+                        int extendedWindowSize = Math.Min(_maxStreamWindowSize, _streamWindowSize * 2);
+                        windowUpdateIncrement += extendedWindowSize - _streamWindowSize;
+                        _streamWindowSize = extendedWindowSize;
+
+                        if (NetEventSource.Log.IsEnabled()) _stream.Trace($"[FlowControl] Updated Stream Window. StreamWindowSize: {StreamWindowSize}, StreamWindowThreshold: {StreamWindowThreshold}");
+
+                        Debug.Assert(_streamWindowSize <= _maxStreamWindowSize);
+                        if (_streamWindowSize == _maxStreamWindowSize)
+                        {
+                            if (NetEventSource.Log.IsEnabled()) _stream.Trace($"[FlowControl] StreamWindowSize reached the configured maximum of {_maxStreamWindowSize}.");
+
+                            // We are at maximum configured window, stop further scaling:
+                            _windowScalingEnabled = false;
+                            _connection._rttEstimator.MaximumWindowReached();
+                        }
+                    }
+                }
+
+                _deliveredBytes = 0;
+
+                Task sendWindowUpdateTask = _connection.SendWindowUpdateAsync(_stream.StreamId, windowUpdateIncrement);
+                _connection.LogExceptions(sendWindowUpdateTask);
+
+                _lastWindowUpdate = currentTime;
+            }
+
+            private static TimeSpan StopwatchTicksToTimeSpan(long stopwatchTicks)
+            {
+                long ticks = (long)(StopWatchToTimesSpan * stopwatchTicks);
+                return new TimeSpan(ticks);
+            }
+        }
+
+        private struct RttEstimator
```

Comment:
I think we should explain the threading assumptions here.

There's one comment down below in MaximumWindowReached, but we really should have a more comprehensive explanation here.